CoreIPCCFDictionary::createCFDictionary() doesn't validate key/value type

pratiksha-99 opened the pull request:
#### 634bc8c5b38a3a90f35b814582349382677740a5
<pre>
CoreIPCCFDictionary::createCFDictionary() doesn&apos;t validate key/value type
<a href="https://bugs.webkit.org/show_bug.cgi?id=289280">https://bugs.webkit.org/show_bug.cgi?id=289280</a>
<a href="https://rdar.apple.com/146384453">rdar://146384453</a>

Reviewed by NOBODY (OOPS!).

Restricted CGColorSpaceSerialization type to hold CoreIPCPlistDictionary values only.

* LayoutTests/ipc/coreipc.js:
(splitClassAndFunction):
(CoreIPCClass.prototype.initializeMessageByName):
(CoreIPCClass.prototype.initializeMessages):
(CoreIPCClass.prototype.generateSendingFunction):
(export.StreamConnection):
(export.StreamConnectionInterface):
(export.StreamConnectionInterface.prototype.getIdentifier):
(export.StreamConnectionInterface.prototype.initializeMessages):
(export.StreamConnectionInterface.prototype.generateStreamSendingFunction):
(export.ArgumentSerializer.parseTemplate):
(export.ArgumentSerializer):
* LayoutTests/ipc/invalid-ipc-cfdictionary-objects-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-ipc-cfdictionary-objects-crash.html: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlist.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.mm:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm:
(WebKit::CoreIPCPlistObject::CoreIPCPlistObject):
(IPC::ArgumentCoder&lt;UniqueRef&lt;WebKit::PlistValue&gt;&gt;::encode):
* Source/WebKit/Shared/cf/CoreIPCCGColorSpace.h:
(WebKit::CoreIPCCGColorSpace::CoreIPCCGColorSpace):
(WebKit::CoreIPCCGColorSpace::toCF const):
* Source/WebKit/Shared/cf/CoreIPCCGColorSpace.serialization.in:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/634bc8c5b38a3a90f35b814582349382677740a5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4860 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100444 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45902 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15287 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23431 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72759 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30024 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98399 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11432 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-controlled-table-row-visibility.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html ... (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86111 "Found 2 new API test failures: TestWebKitAPI.SampledPageTopColor.HitTestHTMLImage, TestWebKitAPI.WebKit.LoadAndDecodeImage (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53090 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11145 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3843 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45238 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82608 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (failure); re-run-api-tests (cancelled)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3938 "Found 23 new test failures: fast/canvas/canvas-createPattern-video-loading.html fast/canvas/canvas-createPattern-video-modify.html fast/images/animated-image-mp4.html fast/images/exif-orientation-background-image-repeat.html fast/images/exif-orientation-svg-feimage.html fast/images/jpegxl-with-color-profile.html fast/text/otsvg-spacing-canvas.html imported/mozilla/svg/filters/feSpecularLighting-1.svg imported/w3c/web-platform-tests/css/css-backgrounds/border-image-repeat-space-011.html imported/w3c/web-platform-tests/css/css-backgrounds/border-image-slice-fill-001.html ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102481 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22447 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16383 "Found 60 new test failures: accessibility/ARIA-reflection.html accessibility/accessibility-crash-focused-element-change.html accessibility/accessibility-crash-setattribute.html accessibility/accessibility-crash-with-dynamic-inline-content.html accessibility/combobox/aria-combobox-control-owns-elements.html accessibility/combobox/aria-combobox-hierarchy.html accessibility/combobox/aria-combobox-no-owns.html accessibility/combobox/mac/aria-combobox-activedescendant.html accessibility/combobox/mac/combobox-activedescendant-notifications.html accessibility/combobox/mac/combobox-inherited-role.html ... (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81774 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22695 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82122 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81146 "Passed tests") | 
| | [❌ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25721 "Found 60 new test failures: http/wpt/webxr/events_transient_pointer_input_source.https.html http/wpt/webxr/xrDevice_requestSession_hand_tracking_feature.https.html http/wpt/webxr/xrDevice_requestSession_previously_enabled_feature_not_requested.https.html http/wpt/webxr/xrFrame_fillJointRadii.html http/wpt/webxr/xrFrame_fillJointRadii_missing_joint_pose.html http/wpt/webxr/xrFrame_fillPoses.html http/wpt/webxr/xrFrame_fillPoses_missing_joint_pose.html http/wpt/webxr/xrFrame_getJointPose.html http/wpt/webxr/xrFrame_getJointPose_missing_joint_pose.html http/wpt/webxr/xrHand.html ... (failure)") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3141 "Found 22 new test failures: fast/canvas/canvas-createPattern-video-loading.html fast/canvas/canvas-createPattern-video-modify.html fast/images/animated-image-mp4.html fast/images/exif-orientation-background-image-repeat.html fast/images/exif-orientation-svg-feimage.html fast/images/jpegxl-with-color-profile.html fast/text/otsvg-spacing-canvas.html imported/mozilla/svg/filters/feSpecularLighting-1.svg imported/w3c/web-platform-tests/css/css-backgrounds/border-image-repeat-space-011.html imported/w3c/web-platform-tests/css/css-backgrounds/border-image-slice-fill-001.html ... (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/15738 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22416 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27555 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22075 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25550 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pratiksha-99 commented on LayoutTests/ipc/coreipc.js:
```diff
@@ -18,18 +18,10 @@ function deepCopy(object) {
     return JSON.parse(JSON.stringify(object));
 }
 
```
Note: These changes to script are from SEAR fuzzing team for testing CoreIPC calls.

rniwa commented on LayoutTests/ipc/invalid-ipc-cfdictionary-objects-crash.html:
```diff
@@ -0,0 +1,120 @@
+    </script>
+
+    <script type="module">
+        if (window.testRunner) {
```
No curly braces around a single line statement.

rniwa commented on Source/WebKit/Shared/cf/CoreIPCCFDictionary.mm:
```diff
@@ -55,6 +55,17 @@
+static bool isValidPropertyListObject(auto object)
+{
+    return [object isKindOfClass:[NSString class]]
+    || [object isKindOfClass:[NSData class]]
```
These should be indented with 4 more spaces like this:
```cpp
    return [object isKindOfClass:[NSString class]]
        || [object isKindOfClass:[NSData class]]
```

pratiksha-99 replied:
But git-webkit linter didn't complain anything.

rniwa replied:
style checker bot isn't perfect.

rniwa commented on Source/WebKit/Shared/cf/CoreIPCCFDictionary.mm:
```diff
@@ -55,6 +55,17 @@
     }];
 }
 
+// Helper function to check if an object is a valid property list object type
```
I don't think this comment is necessary. This information is already conveyed by the function name.

sheeparegreat commented on Source/WebKit/Shared/cf/CoreIPCCFDictionary.mm:
```diff
@@ -64,7 +75,7 @@
         auto key = pair.key.toID();
         auto value = pair.value.toID();
-        if (key && value)
+        if (key && value && isValidPropertyListObject(key) && isValidPropertyListObject(value))
```
Could you add a `[Validator='...']` to the serialization.in?

achristensen07 replied:
I think instead of this we should change KeyValueVector to Vector<KeyValuePair<std::variant<...>, std::variant<...>>>

sheeparegreat replied:
(The effect of this would to push the validation to the deserialization earlier)

achristensen07 reviewed (changes requested):
Is there a reason that we know that all the CFDictionaries serialized across IPC only contain plist types?  Should we do the same to CoreIPCCFArray?

pratiksha-99 committed:
CoreIPCCFDictionary::createCFDictionary() doesn't validate key/value type
https://bugs.webkit.org/show_bug.cgi?id=289280
rdar://146384453

Reviewed by NOBODY (OOPS!).

Restricted CGColorSpaceSerialization type to hold CoreIPCPlistDictionary values only.

* LayoutTests/ipc/coreipc.js:
(splitClassAndFunction):
(CoreIPCClass.prototype.initializeMessageByName):
(CoreIPCClass.prototype.initializeMessages):
(CoreIPCClass.prototype.generateSendingFunction):
(export.StreamConnection):
(export.StreamConnectionInterface):
(export.StreamConnectionInterface.prototype.getIdentifier):
(export.StreamConnectionInterface.prototype.initializeMessages):
(export.StreamConnectionInterface.prototype.generateStreamSendingFunction):
(export.ArgumentSerializer.parseTemplate):
(export.ArgumentSerializer):
* LayoutTests/ipc/invalid-ipc-cfdictionary-objects-crash-expected.txt: Added.
* LayoutTests/ipc/invalid-ipc-cfdictionary-objects-crash.html: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCPlist.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistArray.mm:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistDictionary.mm:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm:
(WebKit::CoreIPCPlistObject::CoreIPCPlistObject):
(IPC::ArgumentCoder<UniqueRef<WebKit::PlistValue>>::encode):
* Source/WebKit/Shared/cf/CoreIPCCGColorSpace.h:
(WebKit::CoreIPCCGColorSpace::CoreIPCCGColorSpace):
(WebKit::CoreIPCCGColorSpace::toCF const):
* Source/WebKit/Shared/cf/CoreIPCCGColorSpace.serialization.in:
